File: 647.回文子串.cpp
class Solution {
public:
    int cnt = 0;

    int countSubstrings(string s) {
        //以第i位为中间，往两边扩充，回文串长度分为奇数和偶数
        for(int i=0; i<s.length(); ++i){
            helper(s, i, i); // 奇数长度
            helper(s, i, i+1); //偶数长度
        }
        return cnt;
    }

    void helper(string s, int start, int end){
        while(start >= 0 && end < s.length() && s[start] == s[end]){
            ++cnt;
            --start;
            ++end;
        }
    }
};

File: 647.回文子串.js
// dp[i][j] 表示从s[i]到s[j]是否是一个回文串
var countSubstrings = function (s) {
    let n = s.length;
    let dp = new Array(n).fill(0).map(() => {
        return new Array(n).fill(false);
    });
    let res = 0;
    for (let i = 0; i < n; ++i) {
        dp[i][i] = true;
        ++res;
    }
    for (let i = n - 2; i >= 0; --i) {
        for (let j = i + 1; j < n; ++j) {
            if (s.charAt(i) === s.charAt(j)) {
                dp[i][j] = (i + 1 <= j - 1) ? dp[i + 1][j - 1] : true;
            }
            if (dp[i][j]) {
                ++res;
            }
        }
    }
    return res;
};


//以第i位为中间，往两边扩充，回文串长度分为奇数和偶数
var countSubstrings = function (s) {
    let n = s.length;
    let res = 0;

    const helper = (s, start, end) => {
        while (start >= 0 && end < n && s.charAt(start) === s.charAt(end)) {
            ++res;
            ++end;
            --start;
        }
    }

    for (let i = 0; i < n; ++i) {
        helper(s, i, i);
        helper(s, i, i + 1);
    }
    return res;
};